File: leetcode/75_LeetCode_Question/Array/maximum_product_subarray/solve.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
            res =  max(nums)
            now_max = now_min = 1
            for num in nums:
                if num == 0:
                    now_min = now_max = 1
                    continue
                tmp_multiplied_max = now_max * num 
                tmp_multiplied_min = now_min*num
                now_max = max(tmp_multiplied_max, tmp_multiplied_min, num)
                now_min = min(tmp_multiplied_max, tmp_multiplied_min, num) 
                res = max(res, now_max)
            return res
if __name__ == "__main__":
    cases = [-2,3,0,-4] # results: 3
    solution = Solution()
    print(solution.maxProduct(cases))

File: leetcode/75_LeetCode_Question/Array/maximum_product_subarray/solve.rs
use std::cmp::{min, max};

fn max_product_subarray(nums: &Vec<i32>) -> i32 {
    let mut max_num: i32 = nums[0] as i32;
    let mut now_max: i32 = 1;
    let mut now_min: i32 = 1;

    for num in nums {
        let now_max_multiplied = now_max *num;
        let now_min_multiplied = now_min * num;
        now_max = max(now_max_multiplied, max(*num, now_min_multiplied));
        now_min = min(now_min_multiplied, min(*num, now_min_multiplied));
        max_num = max(now_max, max_num);
    }
    return max_num;
}

fn main(){
    let test_vec = vec![1,-2,1,4,-2,-1];
    println!("{:?}", max_product_subarray(&test_vec));
}